lib/ansible/plugins/connection/qubes.py

```diff
@@ -0,0 +1,186 @@
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        self._remote_vmname = self._play_context.remote_addr
+        self._connected = False
+        self._hostname = subprocess.check_output(["hostname"]).decode("utf-8").strip()
+        # Default username in Qubes
+        self.user = "user"
+        if self._play_context.remote_user:
+            self.user = self._play_context.remote_user
+
+    def _qubes(self, cmd=None, in_data=None):
+        """
+        run qvm-run executable
+
+        :param cmd: cmd string for remote system
+        :param in_data: data passed to qvm-run-vm's stdin
+        :return: return code, stdout, stderr
+        """
+        display.vvvv("CMD: ", cmd)
+        if not cmd.endswith("\n"):
+            cmd = cmd + "\n"
+        local_cmd = []
+
+        # For dom0
+        local_cmd.extend(["qvm-run", "--pass-io", "--service"])
+        if self.user != "user":
+            # Means we have a remote_user value
+            local_cmd.extend(["-u", self.user])
+
+        local_cmd.append(self._remote_vmname)
+
+        local_cmd.append("qubes.VMShell")
+
+        local_cmd = [to_bytes(i, errors='surrogate_or_strict') for i in local_cmd]
+
+        display.vvvv("Local cmd: ", local_cmd)
+
+        display.vvv("RUN %s" % (local_cmd,), host=self._remote_vmname)
+        p = subprocess.Popen(local_cmd, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # Here we are writing the actual command to the remote bash
+        p.stdin.write(to_bytes("%s" % cmd, errors='surrogate_or_strict'))
+        stdout, stderr = p.communicate(input=in_data)
+        stdout = to_bytes(stdout, errors='surrogate_or_strict')
+        stderr = to_bytes(stderr, errors='surrogate_or_strict')
+        return p.returncode, stdout, stderr
+
+    def _connect(self):
+        """
+        No persistent connection is being maintained.
+        """
+        super(Connection, self)._connect()
+        self._connected = True
+
+    @ensure_connect
+    def exec_command(self, cmd, in_data=None, sudoable=False):
+        """ run specified command in a running QubesVM """
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.vvvv("CMD IS: %s" % cmd)
+
+        rc, stdout, stderr = self._qubes(cmd)
+
+        display.vvvvv("STDOUT %r STDERR %r" % (stderr, stderr))
+        return rc, stdout, stderr
+
+    def put_file(self, in_path, out_path):
+        """ Place a local file located in 'in_path' inside container at 'out_path' """
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self._remote_vmname)
+
+        cmd = "qvm-copy-to-vm " + self._remote_vmname + " " + in_path
+        cmd_args_list = shlex.split(to_native(cmd, errors='surrogate_or_strict'))
```

Comment:
Why construct a string and then split it into a list instead of simply constructing a list?

Reply:
@abadger I just wanted to be sure about that `in_path` value, this looked safe to me. If you say it is okay to just use a list, I don't mind doing that.

Reply:
Yeah, a list is fine.